src/services/pcn-bridge/src/Bridge_dp.c

```diff
@@ -0,0 +1,363 @@
+
+  return 0;
+}
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
+};
+
+BPF_TABLE("hash", u32, struct port, ports, 256);
+
+struct eth_hdr {
+  __be64 dst : 48;
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+static __always_inline int handle_rx(struct CTXTYPE *ctx,
+                                     struct pkt_metadata *md) {
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *eth = data;
+
+  if (data + sizeof(*eth) > data_end)
+    goto DROP;
+
+  u32 in_ifc = md->in_port;
+
+  pcn_log(ctx, LOG_DEBUG, "in_ifc=%d proto=0x%x", in_ifc,
+          bpf_ntohs(eth->proto));
+  pcn_log(ctx, LOG_DEBUG, "mac src:%M dst:%M", eth->src, eth->dst);
+
+  u16 vlanid = 0;
+
+  bool tagged = pcn_is_vlan_present(ctx);
+
+  if (tagged) {
+    int result;
+    if ((result = pcn_get_vlan_id(ctx)) < 0)
+      goto DROP;
+    vlanid = result;
+  }
+
+  struct port *in_port = ports.lookup(&in_ifc);
+  if (!in_port) {
+    pcn_log(ctx, LOG_ERR, "Port %u not found", in_ifc);
+    goto DROP;
+  }
+
+  struct port_vlan_key vlan_key;
+  struct port_vlan_value *vlan_entry;
+
+  vlan_key.port = in_ifc;
+
+  if (in_port->mode == PORT_MODE_ACCESS) {
+    if (tagged)
+      goto DROP;
+
+    vlan_key.vlan = VLAN_WILDCARD;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry) {
+      pcn_log(ctx, LOG_ERR, "no VLAN configured in this access port");
+      goto DROP;
+    }
+
+    vlanid = vlan_entry->vlan;
+
+    pcn_log(ctx, LOG_TRACE, "port mode ACCESS vlan=%d, state=%d", vlanid,
+            vlan_entry->stp_state);
+  } else {
+    // TRUNK
+    if (!tagged) {
+      if (!in_port->native_vlan_enabled) {
+        pcn_log(ctx, LOG_ERR,
+                "Packet is not tagged but the port has native vlan disabled");
+        goto DROP;
+      }
+      vlanid = in_port->native_vlan;
+    }
+    vlan_key.vlan = vlanid;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry) {
+      // TODO manage case of all vlans allowed
+      pcn_log(ctx, LOG_ERR,
+              "VLAN of the packet is not configured in this trunk port");
+      goto DROP;
+    }
+  }
+
+#if STP_ENABLED
+  // TODO: this is probably wrong or could done better, for example adding
+  // a fake entry on the forwarding table
+  // if the packet is a BPDU, send it to the slowpath
+  if (eth->dst == 0x000000C28001 || eth->dst == 0xCDCCCC0C0001) {
+    u32 mdata[3];
+    mdata[0] = vlanid;
+    mdata[1] = tagged;
+    return pcn_pkt_controller_with_metadata(ctx, md, BPDU_REASON, mdata);
+  }
+
+  // Should learning be performed?
+  if (!stp_learn_in_state(vlan_entry->stp_state)) {
+    pcn_log(ctx, LOG_TRACE, "port %d should not learn: state: %d", in_ifc,
+            vlan_entry->stp_state);
+    return RX_DROP;
+  }
+#endif
+
+  u32 now = time_get_sec();
```

Comment:
@mauriciovasquezbernal I would call it "pcn_time_get_sec()" given that it's an helper provided by us.

Reply:
It is internal to the service, so I don't think we should call it `pcn_...`

Reply:
Should't be moved in the library?

Reply:
Yes, it should: https://github.com/polycube-network/polycube/issues/58

Reply:
Considering that this feature will be implemented in the future, i resolve this conversation for the moment.